mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala

```diff
@@ -546,10 +574,88 @@ object KMeans {
       .run(data)
   }
 
+  private[spark] def validateInitMode(initMode: String): Boolean = {
+    initMode match {
+      case KMeans.RANDOM => true
+      case KMeans.K_MEANS_PARALLEL => true
+      case _ => false
+    }
+  }
+  private[spark] def validateDistanceMeasure(distanceMeasure: String): Boolean = {
+    distanceMeasure match {
+      case DistanceMeasure.EUCLIDEAN => true
+      case DistanceMeasure.COSINE => true
+      case _ => false
+    }
+  }
+}
+
+/**
+ * A vector with its norm for fast distance computation.
+ *
+ * @see [[org.apache.spark.mllib.clustering.KMeans#fastSquaredDistance]]
+ */
+private[clustering]
+class VectorWithNorm(val vector: Vector, val norm: Double) extends Serializable {
+
+  def this(vector: Vector) = this(vector, Vectors.norm(vector, 2.0))
+
+  def this(array: Array[Double]) = this(Vectors.dense(array))
+
+  /** Converts the vector to a dense vector. */
+  def toDense: VectorWithNorm = new VectorWithNorm(Vectors.dense(vector.toArray), norm)
+}
+
+
+private[spark] abstract class DistanceMeasure extends Serializable {
+
+  /**
+   * Returns the index of the closest center to the given point, as well as the squared distance.
+   */
+  def findClosest(
+     centers: TraversableOnce[VectorWithNorm],
+     point: VectorWithNorm): (Int, Double)
+
+  /**
+   * Returns the K-means cost of a given point against the given cluster centers.
+   */
+  def pointCost(
+      centers: TraversableOnce[VectorWithNorm],
+      point: VectorWithNorm): Double =
+    findClosest(centers, point)._2
+
+  /**
+   * Returns whether a center converged or not, given the epsilon parameter.
+   */
+  def isCenterConverged(
```

Comment:
Likewise this always seems to be "distance < epsilon"; does it ever vary?

Reply:
It's not for the EuclideanDistance, since in that case for performance reason is: `distance^2 < epsilon^2`.

Anyway, as you suggested for the `findClosest`, I refactored the base class in order to have a default implementation based on the `distance` method I introduced. And then, for the Euclidean distance I am overriding all the methods to the more efficient implementations. This looks the best and cleanest approach to me since it will allow to add more distance measures by implementing only the `distance` method, as the current `CosineDistance` implementation.

Thank you for your comments. Please, when you have time take a look at the new structure and let me know if it looks good to you now.

Thanks.